paddle/fluid/framework/prune.cc

```diff
@@ -269,124 +299,154 @@ void CloneWholeBlock(proto::ProgramDesc* input, proto::ProgramDesc* output,
+  // Step 1. Mark backward, optimize and lrsched ops in the block
+  auto* ops = origin->mutable_ops();
   for (auto op_iter = ops->begin(); op_iter != ops->end(); ++op_iter) {
     auto& op_desc = *op_iter;
-    if (HasFalseTarget(op_desc) ||
-        HasDependentInputVar(op_desc, gradop_output_vars)) {
+    auto op_role = GetOpRole(op_desc);
+    if (op_role & static_cast<int>(OpRole::kOptimize) ||
+        op_role & static_cast<int>(OpRole::kBackward) ||
+        op_role & static_cast<int>(OpRole::kLRSched)) {
       op_desc.set_is_target(false);
-      AppendOpInputVarNames(op_desc, &gradop_input_vars);
-      AppendOpOutputVarNames(op_desc, &gradop_output_vars);
     }
   }
 
-  // Step 4. Mark ops need to be reserved on sub-branch
-  for (auto op_iter = ops->rbegin(); op_iter != ops->rend(); ++op_iter) {
-    auto& op_desc = *op_iter;
-    if (!op_desc.has_is_target()) {
-      if (HasDependentOutputVar(op_desc, gradop_input_vars)) {
-        op_desc.set_is_target(false);
-        AppendOpInputVarNames(op_desc, &gradop_input_vars);
-      } else {
-        op_desc.set_is_target(true);
-        AppendOpInputVarNames(op_desc, &op_input_vars);
-        AppendOpOutputVarNames(op_desc, &op_output_vars);
-      }
-    }
-  }
-
-  // Step 5. Copy the forward ops to new ProgramDesc
-  //   Note: The proto::ProgramDesc doesn't have interface
-  //         to remove op and var
-  auto* op_field = output_block->mutable_ops();
+  // Step 2. Copy the forward ops which have not been set false target to new
+  // ProgramDesc
+  // Note: The proto::ProgramDesc doesn't have interface
+  //       to remove op and var
+  auto* op_field = pruned->mutable_ops();
   op_field->Clear();
   for (auto op_iter = ops->begin(); op_iter != ops->end(); ++op_iter) {
-    if (IsTarget(*op_iter)) {
+    if (!HasFalseTarget(*op_iter)) {
       auto* op = op_field->Add();
+      AppendOpInputVarNames(*op_iter, &op_input_vars);
+      AppendOpOutputVarNames(*op_iter, &op_output_vars);
       *op = *op_iter;
-      if (HasSubBlock(*op)) {
-        CloneWholeBlock(input, output, GetSubBlockIndex(*op), output_block_id);
-      }
     }
   }
 
-  // Step 6. Copy the forward vars to new ProgramDesc
+  // Step 3. Copy the forward vars to new ProgramDesc,
   // construct all var's map before clear
-  auto* var_field = output_block->mutable_vars();
+  auto* origin_vars = origin->mutable_vars();
+  auto* pruned_vars = pruned->mutable_vars();
   std::unordered_map<std::string, proto::VarDesc> var_map;
-  for (const auto& var : *var_field) {
+  for (const auto& var : *origin_vars) {
     var_map[var.name()] = var;
   }
+  pruned_vars->Clear();
+
   std::unordered_set<std::string> var_names;
   var_names.insert(op_input_vars.begin(), op_input_vars.end());
   var_names.insert(op_output_vars.begin(), op_output_vars.end());
-  var_field->Clear();
   for (const auto& name : var_names) {
-    *var_field->Add() = var_map[name];
+    if (var_map.count(name)) {
```

Comment:
Why `name` does not exist in `var_map`? It seems that `var_map` contains all vars from original block and `name` is a var from prune block ops. So, is it better to use `PADDLE_ENFORCE_xxx` to avoid bug?

Reply:
Got it. The var may be from the the parent block or forward block. The code is right here.

Reply:
I add some comments for that.